Lumos walkthrough completed (#69)
CrowdStrike/Cloud-AWS: Container Security/eks-implementation-guide.md

## Pre-requisites

> - Existing AWS Account and VPC

Reviewer: Is a VPC truly needed?  Does this mean LUMOS must be added into an existing VPC? Does one need to be created, and if so, with what permissions?

Author: You're probably right here as eksctl is going to create the VPC for the cluster. Could probably remove this requirement.

> - You will need a workstation with a linux platform

Reviewer: So windows and OSX-based system administrators cannot deploy LUMOS? If that's accurate, can we detail why?

> - Docker installed and running locally on the workstation

Reviewer: Specifically docker (docker.com) or a Linux container engine?

Author: specifically docker is used in these instructions.

> - API Credentials from Falcon with Sensor Download Permissions

Reviewer: Where do the users setup those creds?

  * For this step and practice of least privilege, you would want to create a dedicated API secret and key.
  

> Various command-line utilities are required for this demo. The utilities can either be installed locally or through ready-made tooling container. We recommend the use of the container.

Reviewer: Specifically what command line tools are needed?

## Deployment Configuration Steps

### Step 1: Create EKS Fargate Cluster (For an existing cluster continue to step 2)
 - Set the cloud region (example below uses us-west-1)
   ```$ CLOUD_REGION=us-west-1```
 - Create new EKS Fargate cluster. It may take couple minutes before cluster is fully up and functioning.
   ```
   $ eksctl create cluster \
       --name eks-fargate-cluster --region $CLOUD_REGION \

>        --fargate

Reviewer: +1 to making this a variable!

   ```
   $ eksctl create fargateprofile \
       --region $CLOUD_REGION \
       --cluster eks-fargate-cluster \
       --name fp-falcon-sytem \
       --namespace falcon-system
   ```
   Example output
   ```
   [ℹ]  creating Fargate profile "fp-falcon-sytem" on EKS cluster "eks-fargate-cluster"
   [ℹ]  created Fargate profile "fp-falcon-sytem" on EKS cluster "eks-fargate-cluster"
   ```

 - Verify that your local kubectl utility has been configured to connect to the cluster.
   ```
   $ kubectl cluster-info
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
   Kubernetes control plane is running at https://EEAB38XXXXXXXXXXXXXXXXXXXXXXXXXX.sk1.eu-west-1.eks.amazonaws.com
   CoreDNS is running at https://EEAB38XXXXXXXXXXXXXXXXXXXXXXXXXX.sk1.eu-west-1.eks.amazonaws.com/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy

   To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
   ```
   kubectl is command line tool that lets you control Kubernetes clusters. For configuration, kubectl looks for a file named config in the `$HOME/.kube` directory. This config was created previously by `eksctl create cluster` command and contains login information for your newly created cluster.

### Step 2: Create Container Repository (For an existing ECR continue to step 3)

 - Create container repository in AWS ECR. AWS ECR (Elastic Container Registry) is a cloud service providing a container registry. The bellow command creates a new repository in the registry and this repository will be subsequently used to store the Falcon Container sensor image.
   ```
   $ aws ecr create-repository --region $CLOUD_REGION --repository-name falcon-sensor
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   }
   ```

 - Note the `repositoryUri` of the newly created repository to the environment variable for further use. Falcon Container Sensor will be available under this URI.

### Step 3: Push the falcon sensor image to the Repository

 - Set the FALCON_IMAGE_URI variable to your managed ECR based on the ECR `repositoryUri`
 
 - Note for existing ECR registries the `repositoryUri` can be found with the following command.
   ```
   aws ecr describe-repositories
   ```

>    Example

Reviewer: ```suggestion
   Example output:
```

Author: This one is actually an example. I will add the colon for consistency.

    ```
   $ FALCON_IMAGE_URI=123456789123.dkr.ecr.eu-west-1.amazonaws.com/falcon-sensor:latest
   ```

>    - Note that the URI suffix `:latest` is a container image tag. Repository usually contains multiple image version, and tags are used to distinguish between versions. This must be added to the end of the `repositoryUri`.

Reviewer: ```suggestion
   - Note that the URI suffix `:latest` is a container image tag. Container repositories usually contain multiple image versions, and tags are used to distinguish between versions. A container image tag must be added to the end of the `repositoryUri`.
```

 - Obtain the falcon-sensor container tarball.
 - Note that the prompts will require the input of the API key and secret that were obtained in the pre-requisites
   ```
   $ falcon_sensor_download --os-name=Container
   ```

>    Example prompts

Reviewer: ```suggestion
   Example prompts:
```

   ```
   Please provide your Falcon Client ID: ABC
   Please provide your Falcon Client Secret: XYZ
   Downloaded Falcon Usermode Container Sensor to falcon-sensor-6.21.0-403.container.x86_64.tar.bz2
   ```
 - Import the tarball to your local docker. If you are following this guide inside the tooling
   container, you can run this command outside of the container as the docker socket is shared
   between your host system and the said tooling container.
   ```
   $ docker load -i falcon-sensor-6.21.0-403.container.x86_64.tar.bz2
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
   30cbe59c0010: Loading layer  39.07MB/39.07MB
   Loaded image: falcon-sensor-6.21.0-403.container.x86_64.tar.bz2
   ```
 - List the loaded docker image
   ```
   docker images | grep falcon-sensor
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
   falcon-sensor    6.21.0-403.container.x86_64.Release.US2   xxxxxxx        2 weeks ago         78.8MB
   ```
 - Note the image name and the image tag in the first and second columns respectively
 - Using the local image name and tag, re-tag the image for your managed ECR using the variable previously set
   ```
   $ docker tag falcon-sensor:6.21.0-403.container.x86_64.Release.US2 $FALCON_IMAGE_URI
   ```
 - Push the image to AWS ECR
   ```
   $ docker push $FALCON_IMAGE_URI
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   $ docker run --rm --entrypoint installer $FALCON_IMAGE_URI \
       -cid $CID -image $FALCON_IMAGE_URI \
       | kubectl apply -f -
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

 - (optional) Watch the progress of a deployment
   ```
   $ watch 'kubectl get pods -n falcon-system'
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

 - Instruct Kubernetes cluster to start a detection application
   ```
   $ kubectl apply -f ~/demo-yamls/detection-single.yaml
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
   deployment.apps/detection-single created
   ```
 - (optional) See the logs of the admission installer to ensure it is responding to the detection app start-up
   ```
   $ kubectl logs -n falcon-system $(kubectl get pods -n falcon-system | awk 'FNR > 1' | awk '{print $1}')
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
   injector server starting ...
   2021/02/03 16:05:51 Handling webhook request with id 0d20df1d-8737-4bf0-bea6-fd03b48b2516 in namespace default ...
   2021/02/03 16:05:51 Webhook request with id 0d20df1d-8737-4bf0-bea6-fd03b48b2516 in namespace default handled successfully!
   ```
 - (optional) Watch the deployment progress of the detection app
   ```
   $ watch 'kubectl get pods'
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
   NAME                            READY   STATUS    RESTARTS   AGE
   detection-single-767cd557b-267zg   2/2     Running   0          2m26s
   ```
 - (optional) Ensure that the newly created pod was allocated an Agent ID (AID) from CrowdStrike Falcon platform
   ```
   $ kubectl exec $(kubectl get pods | grep detection | awk '{print $1}') -c falcon-container -- falconctl -g --aid
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   $ docker run --rm --entrypoint installer $FALCON_IMAGE_URI \
       -cid $CID -image $FALCON_IMAGE_URI \
       | kubectl delete -f -
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
 - Step 3: Delete the falcon image from AWS ECR registry
   ```
   $ aws ecr batch-delete-image --region $CLOUD_REGION \
       --repository-name falcon-sensor \
       --image-ids imageTag=latest
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
 - Step 4: Delete the AWS ECR repository
   ```
   $ aws ecr delete-repository --region $CLOUD_REGION --repository-name falcon-sensor
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```

   ```
 - Step 5: Delete the AWS EKS Fargate Cluster
   ```
   $ eksctl delete cluster --region $CLOUD_REGION eks-fargate-cluster
   ```

>    Example output

Reviewer: ```suggestion
   Example output:
```
